pulsar-broker/src/main/java/org/apache/pulsar/compaction/TwoPhaseCompactor.java

```diff
@@ -217,10 +229,11 @@ private void phaseTwoLoop(RawReader reader, MessageId to, Map<String, MessageId>
                         }
                     } else {
                         Pair<String,Integer> keyAndSize = extractKeyAndSize(m);
+                        MessageId msg;
                         if (keyAndSize == null) { // pass through messages without a key
                             messageToAdd = Optional.of(m);
-                        } else if (latestForKey.get(keyAndSize.getLeft()).equals(id)
-                                   && keyAndSize.getRight() > 0) {
+                        } else if ((msg = latestForKey.get(keyAndSize.getLeft())) != null && msg.equals(id)
+                                && keyAndSize.getRight() > 0) { // consider message only if present into latestForKey map
```

Comment:
I guess this check is redundant now. No harm in leaving it though

Reply:
that's correct. it's redundant so, I think I will remove it as well.

Reply:
change it to a checkArgument()